group/assignment6/4th_challenge.py

```diff
@@ -0,0 +1,68 @@
+from collections import defaultdict
+from itertools import permutations
+
+
+class Graph:
+    """
+    Args:
+        nums_to_states: {num_of_state: state}
+        states_to_nums: {state: num_of_state}
+    """
+    def __init__(self):
+        self.vertices = set()
+        self.edges = defaultdict(list)
+        self.nums_to_states = dict()
+        self.states_to_nums = dict()
+
+    def print_all_paths(self, start, end, visited=None, path=None, all_paths=None):
+        if visited is None:
+            visited = dict.fromkeys(self.vertices, False)
+        if path is None:
+            path = []
+        if all_paths is None:
+            all_paths = []
+        visited[start] = True
+        path.append(start)
+        
+        if start == end and len(path) != 1:
+            all_paths.append([self.nums_to_states[x] for x in path])
+        else:
+            for i in self.edges[start]:
+                if not visited[i]:
+                    self.print_all_paths(i, end, visited, path, all_paths)
+
+        path.pop()
+        visited[end] = False
```

Comment:
You actually never update visited[end] = True, there is a bug here. Try any example of size n=4, the number of paths you found will be too small.

Reply:
@aliskin 24 row visited[start]=True?